Source/SwiftLintFramework/Rules/Style/RedundantDiscardableLetRule.swift

```diff
@@ -29,62 +28,65 @@ public struct RedundantDiscardableLetRule: SubstitutionCorrectableRule, Configur
+            }
         }
     }
 
-    private func isInBooleanCondition(byteOffset: ByteCount, dictionary: SourceKittenDictionary) -> Bool {
-        guard let byteRange = dictionary.byteRange, byteRange.contains(byteOffset) else {
-            return false
-        }
+    final class Rewriter: SyntaxRewriter, ViolationsSyntaxRewriter {
+        private(set) var correctionPositions: [AbsolutePosition] = []
+        let locationConverter: SourceLocationConverter
+        let disabledRegions: [SourceRange]
 
-        let kinds: Set<StatementKind> = [.if, .guard, .while]
-        if let kind = dictionary.statementKind, kinds.contains(kind) {
-            let conditionKind = "source.lang.swift.structure.elem.condition_expr"
-            for element in dictionary.elements where element.kind == conditionKind {
-                guard let elementRange = element.byteRange, elementRange.contains(byteOffset) else {
-                    continue
-                }
+        init(locationConverter: SourceLocationConverter, disabledRegions: [SourceRange]) {
+            self.locationConverter = locationConverter
+            self.disabledRegions = disabledRegions
+        }
 
-                return true
+        override func visit(_ node: VariableDeclSyntax) -> DeclSyntax {
+            guard node.hasRedundantDiscardableLetViolation, !isInDisabledRegion(node) else {
+                return super.visit(node)
             }
-        }
 
-        for subDict in dictionary.substructure where
-            isInBooleanCondition(byteOffset: byteOffset, dictionary: subDict) {
-                return true
+            correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
+            let newNode = node
+                .withLetOrVarKeyword(nil)
+                .withBindings(node.bindings.withLeadingTrivia(node.letOrVarKeyword.leadingTrivia))
+            return super.visit(newNode)
         }
 
-        return false
+        private func isInDisabledRegion<T: SyntaxProtocol>(_ node: T) -> Bool {
+            disabledRegions.contains { region in
+                region.contains(node.positionAfterSkippingLeadingTrivia, locationConverter: locationConverter)
+            }
+        }
     }
+}
 
-    private func hasExplicitType(utf16Range: Range<Int>, fileContents: NSString) -> Bool {
-        guard utf16Range.upperBound != fileContents.length else {
-            return false
-        }
-        let nextUTF16Unit = fileContents.substring(with: NSRange(location: utf16Range.upperBound, length: 1))
-        return nextUTF16Unit == ":"
+private extension VariableDeclSyntax {
+    var hasRedundantDiscardableLetViolation: Bool {
+        letOrVarKeyword.tokenKind == .letKeyword &&
+            bindings.count == 1 &&
+            bindings.first!.pattern.is(WildcardPatternSyntax.self) &&
+            bindings.first!.typeAnnotation == nil &&
+            modifiers?.contains(where: { $0.name.text == "async" }) != true
```

Comment:
```suggestion
            modifiers?.contains(where: { $0.name.text == "async" }) == false
```

Reply:
We want `nil` and `false` to both satisfy the guard.

Another way to write this would be:

```suggestion
            modifiers == nil || !modifiers.first!.contains(where: { $0.name.text == "async" })
```

But that's significantly harder to read.

Reply:
I see. Optional booleans ... 😅